RingBuffer: Add iterator() and change peek() to return a pointer

jorangreef opened the pull request:
We can now iterate ring buffer items as pointers, and with `peek()` now returning a pointer we can also mutate the peeked item.

The reason for adding the iterator is for use within `Replica`, to iterate a ring buffer of inflight prepares.

jorangreef commented on src/client.zig:
```diff
@@ -181,7 +181,7 @@ pub const Client = struct {
 
     fn on_request_timeout(self: *Client) void {
-        const current_request = self.request_queue.peek() orelse return;
+        const current_request = (self.request_queue.peek() orelse return).*;
```
Apologies for introducing this blemish. Couldn't figure a better way to do it.

jorangreef replied:
Ah of course.

ifreund commented on src/client.zig:
```diff
@@ -239,7 +239,7 @@ pub const Client = struct {
 
         if (self.request_queue.peek()) |next_request| {
-            self.send_request(next_request.message);
+            self.send_request(next_request.*.message);
```
This isn't necessary to access a field of the pointed to struct, the semantics are exactly the same as the status quo code.
```suggestion
            self.send_request(next_request.message);
```

ifreund commented on src/message_bus.zig:
```diff
@@ -996,7 +996,7 @@ fn MessageBusImpl(comptime process_type: ProcessType) type {
                 assert(self.state == .connected);
                 assert(self.fd != -1);
-                const message = self.send_queue.peek() orelse return;
+                const message = (self.send_queue.peek() orelse return).*;
```
Here we do have to dereference the result as `peek()` returns a `**Message`, so this change is necessary albeit a little ugly.

ifreund commented on src/ring_buffer.zig:
```diff
@@ -20,54 +23,74 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
-            self.index = (self.index + 1) % self.buffer.len;
-            self.count -= 1;
-            return ret;
+            defer {
+                self.index = (self.index + 1) % self.buffer.len;
+                self.count -= 1;
+            }
+            return self.buffer[self.index];
```
This is sweet :)

jorangreef replied:
Cool :) Glad you agree. I was in two minds about it, since it's not resource deallocation and it's upside down... but we save coming up with a temp local scope variable name 😄

ifreund commented on src/ring_buffer.zig:
```diff
@@ -20,54 +23,74 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
         /// Return but do not remove the next item, if any.
-        pub fn peek(self: *Self) ?T {
+        pub fn peek(self: *Self) ?*T {
             if (self.empty()) return null;
-            return self.buffer[self.index];
+            return &self.buffer[self.index];
         }
```
One way to eliminate the pain this change causes when using a ring buffer of pointers such as the send queue in the message bus would be to keep this function as-is and instead add a `peek_ptr()` variant that returns `?*T` which could be used when copying is not desired. The same could be done for the `iterator()` function. For the client code where we are storing struct directly in the ring buffer's array instead of pointers, using the `*_ptr()` variants would avoid copies.

What do you think about that?

jorangreef replied:
Yes, agreed. I was actually also thinking we need some way to peek/iterate a const value (via a copy that can't reach back) or pointer, in the same way that you can iterate over a slice, capturing by const value or by pointer.

ifreund commented on src/ring_buffer.zig:
```diff
@@ -20,54 +23,74 @@ pub fn RingBuffer(comptime T: type, comptime size: usize) type {
+        };
+
+        /// Returns an iterator to iterate through all `count` items in the ring buffer.
+        pub fn iterator(self: *Self) Iterator {
```
This could probably use a comment stating that `push()`/`pop()` while iterating are not safe.

jorangreef commented:
Thanks for the review. I added `peek_ptr()` and `next_ptr()`, and used these as the underlying implementation for `peek()` and `next()` so that our test coverage tests both simultaneously. I also updated Benchmark and Client to access by ptr and to use if-unwrapping rather than introduce a temporary variable in the scope.

jorangreef commented:
Merging so long, please let me know if I can change anything further.
